### PR TITLE
Add episode 24

### PR DESCRIPTION
**Pull request by bitsondatadev**

### CODE REVIEW

File: _episodes/24.md [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+   - Node - The physical machine or VM running in the kubernetes cluster. 
+   - Cluster - a set of nodes connected to the same control plane node.
+   - POD - single instance of an application, the smallest object in kubernetes.
+
+   <p align="center">
+    <img align="center" width="75%" height="100%" src="/assets/episode/24/components-of-kubernetes.svg"/><br/>
+    Source: https://kubernetes.io/docs/concepts/overview/components/
+   </p>
+
+   Kubernetes control plane consists of:
```
[the remark this answers is not in the document]
Comment (mosabua) [reply _episodes/24.md:92]: same for other lists below

File: _episodes/24.md [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+his Trino helm chart so that it can be maintained in the community version.
+
+Valeriano Trino Helm Chart: <https://artifacthub.io/packages/helm/valeriano-manassero/trino>
+
+It hasn't been merged yet but we are really looking forward to seeing this get
+merged in. Thanks Valeriano!
+
+## Demo: Running the Trino charts with kubectl
+
+For this weeks Demo, you will need to install [kubectl](https://kubernetes.io/docs/tasks/tools/),
```
Comment (mosabua) [inline _episodes/24.md:158]: demo, you need to

File: _episodes/24.md [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+[minikube](https://minikube.sigs.k8s.io/docs/start/) using the [docker driver](https://minikube.sigs.k8s.io/docs/drivers/docker/),
+and [helm](https://helm.sh/docs/intro/install/). You will also be referencing
```
Comment (mosabua) [inline _episodes/24.md:160]: Reword to avoid future tense

File: _episodes/24.md [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+## Demo: Running the Trino charts with kubectl
+
+For this weeks Demo, you will need to install [kubectl](https://kubernetes.io/docs/tasks/tools/),
+[minikube](https://minikube.sigs.k8s.io/docs/start/) using the [docker driver](https://minikube.sigs.k8s.io/docs/drivers/docker/),
+and [helm](https://helm.sh/docs/intro/install/). You will also be referencing 
+the ArtifactHub URL for this.
+
+https://artifacthub.io/packages/helm/trino/trino
+
+```
```
Comment (mosabua) [inline _episodes/24.md:165]: Break all this below up into multiple code snippets and text

**Review by mosabua — CHANGES_REQUESTED** [review]
Needs a bunch of rework and simplification on wording. Avoid future and past tense and would/could/should and use consistent lists. Ping me for another look when you went over it again.

**Comment by bitsondatadev** [comment]
@mosabua revised

File: _episodes/24.md [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+---
+
+This is the first episode in a series where we cover the basics and just enough
+advanced Kubernetes features and information to understand how to deploy Trino 
+on Kubernetes.
+
+## Concept of the week: K8s architecture: Containers, Pods, and kubelets
+
+For this concept of the week, we want to provide you a minimalistic overview of
+what you need to know for Kubernetes (also abbreviated k8s) to deploy a simple
```
Comment (mosabua) [inline _episodes/24.md:27]: to know about Kubernetes to deploy Trino to a cluster.

File: _episodes/24.md [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+   you to indicate how to manage containers declaritively using yaml 
+   configuration files. This definition can be tricky to understand if you don't
+   have proper context. To make sure nobody is left behind, it is useful to 
+   cover what containers are.
+    
+   - The traditional way to deploy an application, is to take the compiled 
+   binary of that application and run it directly on computer hardware that has
+   an operating system to run the application on it. This works, but had a lot
+   of dependency on the underlying hardware an operating system to be functional
+   and required multiple applications to share the same resources. If one of the
```
Comment (mosabua) [inline _episodes/24.md:40]: requires

File: _episodes/24.md [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+   applications fails and cause any of the shared resources to crash, it could
```
Comment (mosabua) [inline _episodes/24.md:41]: and causes

File: _episodes/24.md [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+   affect all applications to fail on that machine. 
+    
+   - To remove these dependencies, engineers created virtual machines (VMs) by 
+   using a VM manager called the hypervisor that could run virtual environments
```
Comment (mosabua) [inline _episodes/24.md:45]: that emulate hardware environments to host other operating systems.

File: _episodes/24.md [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+   an operating system to run the application on it. This works, but had a lot
+   of dependency on the underlying hardware an operating system to be functional
+   and required multiple applications to share the same resources. If one of the
+   applications fails and cause any of the shared resources to crash, it could
+   affect all applications to fail on that machine. 
+    
+   - To remove these dependencies, engineers created virtual machines (VMs) by 
+   using a VM manager called the hypervisor that could run virtual environments 
+   that hosted other operating systems. This was a big step forward as now each 
+   application could be isolated, but it came at a great cost. Each virtual machine
```
Comment (mosabua) [inline _episodes/24.md:47]: This is  ... can be isolated, but it comes at a

File: _episodes/24.md [hunk 10]
```diff
@@ -0,0 +1,413 @@
...
+   hosted an entire operating system and was resource intensive and slow.
```
Comment (mosabua) [inline _episodes/24.md:48]: hosts. .. and is

File: _episodes/24.md [hunk 11]
```diff
@@ -0,0 +1,413 @@
...
+
+   - Containers are the final type of deployment. Containers enable a logical
```
Comment (mosabua) [inline _episodes/24.md:50]: are the newest type

File: _episodes/24.md [hunk 12]
```diff
@@ -0,0 +1,413 @@
...
+
+   <p align="center">
+    <img align="center" width="75%" height="100%" src="/assets/episode/24/container-evolution.svg"/><br/>
+    Source: https://kubernetes.io/docs/concepts/overview/what-is-kubernetes/
+   </p>
+   
+   - Containerization simplified a trend in service architecture called 
+   microservices. Microservices deploy loosely coupled and modular applications
+   rather than all-encompassing monolithic applications. With containers, these
+   applications can be deployed and scale up quickly across various virtual and
```
Comment (mosabua) [inline _episodes/24.md:66]: and scaled up

File: _episodes/24.md [hunk 13]
```diff
@@ -0,0 +1,413 @@
...
+   physical machines without affecting applications on the same machine. This
+   was great, but new problems emerged, such as, monitoring the health of
```
Comment (mosabua) [inline _episodes/24.md:68]: is great, but results in new complexities. Example are the need for new approaches to ..

File: _episodes/24.md [hunk 14]
```diff
@@ -0,0 +1,413 @@
...
+   and this is exactly what Kubernetes solves!
+   
+   <p align="center">
+    <img align="center" width="75%" height="100%" src="/assets/episode/24/load-balancer.jpeg"/><br/>
+    Source: https://www.slideshare.net/devopsdaysaustin/continuously-delivering-microservices-in-kubernetes-using-jenkins<br/>
+    Here we hae two services that each sit behind a load balancer provided and mapped by the Kuberenets cluster.
+   </p>
+
+1. **Kubernetes components and architecture**:
+   - Node - The physical machine or VM running a kubelet and container runtime.
```
Comment (mosabua) [inline _episodes/24.md:81]: need to define what kubelet is before this  .. or avoid using it here

File: _episodes/24.md [hunk 15]
```diff
@@ -0,0 +1,413 @@
...
+   
+   <p align="center">
+    <img align="center" width="75%" height="100%" src="/assets/episode/24/load-balancer.jpeg"/><br/>
+    Source: https://www.slideshare.net/devopsdaysaustin/continuously-delivering-microservices-in-kubernetes-using-jenkins<br/>
+    Here we hae two services that each sit behind a load balancer provided and mapped by the Kuberenets cluster.
+   </p>
+
+1. **Kubernetes components and architecture**:
+   - Node - The physical machine or VM running a kubelet and container runtime. 
+   - Cluster - a set of nodes connected to the same control plane node.
```
Comment (mosabua) [inline _episodes/24.md:82]: what is control plan?

File: _episodes/24.md [hunk 16]
```diff
@@ -0,0 +1,413 @@
...
+   - Pod - single instance of an application, the smallest object in kubernetes.
+
+   <p align="center">
+    <img align="center" width="75%" height="100%" src="/assets/episode/24/components-of-kubernetes.svg"/><br/>
+    Source: https://kubernetes.io/docs/concepts/overview/components/
+   </p>
+
+   Kubernetes control plane components:
+     - API server that kubelets connect with and is the front end for
```
Comment (mosabua) [inline _episodes/24.md:91]: empty line above list and dont indent the list items.

Comment (mosabua) [reply _episodes/24.md:91]: also all items in this list probably should be full sentences.

File: _episodes/24.md [hunk 17]
```diff
@@ -0,0 +1,413 @@
...
+     - API server that kubelets connect with and is the front end for 
+     users and administrators of the cluster.
+     - etcd keystore is a distributed store containing all data used to manage 
+     the cluster
+     - Scheduler that distributes work across nodes and assigns newly created 
+     containers to nodes
+     - Controllers that are the brain behind orchestration and monitors for 
+     nodes going down etc...
+     
+   Kubernetes worker nodes components:
```
Comment (mosabua) [inline _episodes/24.md:100]: empty line before list and dont indent items

File: _episodes/24.md [hunk 18]
```diff
@@ -0,0 +1,413 @@
...
+     - container runtime - is the underlying runtime used to manage containers 
+     - kubelet - agent that checks the health and manages the pods running on the node based on the desired state provided in the PodSpec
+     - kube-proxy - network proxy that maintains network rules applied to nodes and allows network access between Pods in a cluster
+ 
+   You can scale up multiple PODS on a single node until the node has no more
```
Comment (mosabua) [inline _episodes/24.md:105]: pods

File: _episodes/24.md [hunk 19]
```diff
@@ -0,0 +1,413 @@
...
+     - Controllers that are the brain behind orchestration and monitors for 
+     nodes going down etc...
+     
+   Kubernetes worker nodes components:
+     - container runtime - is the underlying runtime used to manage containers 
+     - kubelet - agent that checks the health and manages the pods running on the node based on the desired state provided in the PodSpec
+     - kube-proxy - network proxy that maintains network rules applied to nodes and allows network access between Pods in a cluster
+ 
+   You can scale up multiple PODS on a single node until the node has no more 
+   resources, at which time a new node will need to be added and pod instances will
```
Comment (mosabua) [inline _episodes/24.md:106]: needs to be added .. are distrbuted

File: _episodes/24.md [hunk 20]
```diff
@@ -0,0 +1,413 @@
...
+   be distributed between the nodes. 
+
+1. **So how does this relate to Trino?** 
+     - Why not just use docker compose? Docker-compose is a tool that creates
```
Comment (mosabua) [inline _episodes/24.md:110]: Indent left

File: _episodes/24.md [hunk 21]
```diff
@@ -0,0 +1,413 @@
...
+     multiple containers on a single host, where kubernetes is a platform that
+     allows complex distributed workloads and self healing applications across
+     nodes.
+     - Out of the box for any Kubernetes cluster can do these key things for Trino.
```
Comment (mosabua) [inline _episodes/24.md:114]: ? what does this sentence mean?

File: _episodes/24.md (hunk 20 above)
Comment (mosabua) [reply _episodes/24.md:110]: also make this simpler .. or even remove docker-compose altogether ... 

If you want to keep it just use something like

- Why not just use docker-compose? Docker compose is only suitable for managing multi containers on one host. It has no orchestration features and does not support multiple hosts.

File: _episodes/24.md [hunk 22]
```diff
@@ -0,0 +1,413 @@
...
+        - Simple scale up and down (manually tell k8s to startup or kill).
+        - Kubernetes supports failover, meaning that your workers will self heal if they die.
```
Comment (mosabua) [inline _episodes/24.md:116]: dont talk about self heal. .. Just use "restart"

File: _episodes/24.md [hunk 23]
```diff
@@ -0,0 +1,413 @@
...
+     - Advanced jobs that could exist but not currently in open source.
+        - Auto-scaling via the [Horizontal Pod Autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) 
+        and custom metrics.
+        - Graceful Shutdowns are hooks that you can add into your cluster that
```
Comment (mosabua) [inline _episodes/24.md:120]: something is missing here?

File: _episodes/24.md [hunk 24]
```diff
@@ -0,0 +1,413 @@
...
+         <p align="center">
+            <img align="center" width="75%" height="100%" src="/assets/episode/24/kubernetes-shutdown.svg"/><br/>
+            Source: https://learnk8s.io/graceful-shutdown
+         </p>
+         <p align="center">
+            <img align="center" width="75%" height="100%" src="/assets/episode/24/graceful-shutdown.svg"/><br/>
+            Source: https://learnk8s.io/graceful-shutdown
+         </p>
+1. **What the heck are helm charts then?**
```
Comment (mosabua) [inline _episodes/24.md:129]: Empty line after title

File: _episodes/24.md [hunk 25]
```diff
@@ -0,0 +1,413 @@
...
+According to the helm site, Helm is the best way to find, share, and use
```
Comment (mosabua) [inline _episodes/24.md:130]: Rejig this paragraph logically

- Helm is package manager for K8s
- Removes the need for managing lots of k8s related yaml files
- Best way to deploy apps to k8s
- Charts are available for many different applications
- Helm chart for Trino

File: _episodes/24.md [hunk 26]
```diff
@@ -0,0 +1,413 @@
...
+ 
+Fun Fact: Because Trino doesn't store data it becomes a good candidate 
+application to run on Kubernetes. However, at the end of the day we rely on 
+databases that must exist outside, so it is only so useful.
+
+## PR of the week: PR 11 Merge contributor version of k8s charts with the community version
+
+This weeks [PR of the week](https://github.com/trinodb/charts/pull/11) comes 
+from a different repo under the trinodb org, [trinodb/charts](https://github.com/trinodb/charts).
+This PR contains the merging from contributer [Valeriano Manassero](https://github.com/valeriano-manassero).
```
Comment (mosabua) [inline _episodes/24.md:146]: contributor

File: _episodes/24.md [hunk 27]
```diff
@@ -0,0 +1,413 @@
...
+his Trino helm chart so that it can be maintained in the community version.
+
+Valeriano Trino Helm Chart: <https://artifacthub.io/packages/helm/valeriano-manassero/trino>
+
+It hasn't been merged yet but we are really looking forward to seeing this get
+merged in. Thanks Valeriano!
+
+## Demo: Running the Trino charts with kubectl
+
+For this weeks demo, you will need to install [kubectl](https://kubernetes.io/docs/tasks/tools/),
```
Comment (mosabua) [inline _episodes/24.md:160]: you need to

**Review by mosabua — CHANGES_REQUESTED** [review]

**Review by mosabua — APPROVED** [review]
Just the contributer needs fixing .. then good to go

**Comment by hashhar** [comment]
Unrelated to this PR itself but we might want to link to the channel directly here - 
image

File: _episodes/24.md [hunk 28]
```diff
@@ -0,0 +1,412 @@
...
+ - API server that nodes connect with and is the front end for users and 
+ administrators of the cluster.
+ - etcd keystore is a distributed store containing all data used to manage 
+ the cluster
+ - Scheduler that distributes work across nodes and assigns newly created 
+ containers to nodes
+ - Controllers that are the brain behind orchestration and monitors for 
+ nodes going down etc...
+     
+Kubernetes worker nodes components:
```
Comment (hashhar) [inline _episodes/24.md:104]:
```suggestion
Kubernetes worker node components:
```

File: _episodes/24.md [hunk 29]
```diff
@@ -0,0 +1,412 @@
...
+ - container runtime - is the underlying runtime used to manage containers
```
Comment (hashhar) [inline _episodes/24.md:105]: for consistency with other bullets below

```suggestion
 - container runtime - underlying runtime used to manage containers 
```

File: _episodes/24.md [hunk 30]
```diff
@@ -0,0 +1,412 @@
...
+ - etcd keystore is a distributed store containing all data used to manage 
+ the cluster
+ - Scheduler that distributes work across nodes and assigns newly created 
+ containers to nodes
+ - Controllers that are the brain behind orchestration and monitors for 
+ nodes going down etc...
+     
+Kubernetes worker nodes components:
+ - container runtime - is the underlying runtime used to manage containers 
+ - kubelet - agent that checks the health and manages the pods running on the node based on the desired state provided in the PodSpec
```
Comment (hashhar) [inline _episodes/24.md:106]: I don't think PodSpec has been introduced before this point? No idea if it matters or not.

File: _episodes/24.md [hunk 31]
```diff
@@ -0,0 +1,412 @@
...
+ - kube-proxy - network proxy that maintains network rules applied to nodes and allows network access between Pods in a cluster
+
+   You can scale up multiple pods on a single node until the node has no more 
+   resources, at which time a new node needs to be added and pod instances are 
+   distributed between the nodes. 
+
+1. **So how does this relate to Trino?**
```
Comment (hashhar) [inline _episodes/24.md:113]: Should be a heading maybe - gets kind of lost in the final page.

image

File: _episodes/24.md [hunk 32]
```diff
@@ -0,0 +1,412 @@
...
+
+ - Out of the box, Kubernetes can do these key things for Trino.
+    - Simple scale up and down (manually tell k8s to startup or kill).
```
Comment (hashhar) [inline _episodes/24.md:116]: WDYT?

```suggestion
    - Simple scale up and down (manually tell k8s to start or kill Trino pods).
```

File: _episodes/24.md [hunk 33]
```diff
@@ -0,0 +1,412 @@
...
+     <p align="center">
+        <img align="center" width="75%" height="100%" src="/assets/episode/24/kubernetes-shutdown.svg"/><br/>
+        Source: https://learnk8s.io/graceful-shutdown
+     </p>
+     <p align="center">
+        <img align="center" width="75%" height="100%" src="/assets/episode/24/graceful-shutdown.svg"/><br/>
+        Source: https://learnk8s.io/graceful-shutdown
+     </p>
+     
+1. **What the heck are helm charts then?**
```
Comment (hashhar) [inline _episodes/24.md:132]: Maybe a heading?

File: _episodes/24.md [hunk 34]
```diff
@@ -0,0 +1,412 @@
...
+This weeks [PR of the week](https://github.com/trinodb/charts/pull/11) comes 
+from a different repo under the trinodb org, [trinodb/charts](https://github.com/trinodb/charts).
+This PR contains the merging from contributor [Valeriano Manassero](https://github.com/valeriano-manassero). 
+
+Valerino maintains a [very useful helm chart](https://github.com/valeriano-manassero/helm-charts/tree/main/valeriano-manassero/trino), 
+that started before the Trino org had defined our own community chart. This pull
+request effectively is trying to merge some useful features Valeriano added to 
+his Trino helm chart so that it can be maintained in the community version.
+
+Valeriano Trino Helm Chart: <https://artifacthub.io/packages/helm/valeriano-manassero/trino>
```
Comment (hashhar) [inline _episodes/24.md:152]:
```suggestion
Valeriano's Trino Helm Chart: <https://artifacthub.io/packages/helm/valeriano-manassero/trino>
```

**Review by hashhar — COMMENTED** [review]
Some comments.

I've yet to try the demo to verify if it works for me.

File: _episodes/24.md (hunk 30 above)
Comment (bitsondatadev) [reply _episodes/24.md:106]: If anything hasn't been covered we point to the kubernetes documentation

File: _episodes/24.md (hunk 33 above)
Comment (bitsondatadev) [reply _episodes/24.md:132]: Te be consistent with previous i'll keep this the same

**Review by hashhar — APPROVED** [review]

File: _episodes/24.md [hunk 35]
```diff
@@ -0,0 +1,412 @@
...
+kubectl delete service tcb-trino
+kubectl delete deployment tcb-trino-coordinator
+kubectl delete configmap --all
+```
+
+Now you can run the same demo using the helm chart which includes all of these
+templates out-of-the-box. First add the trino helm chart, check the templates
+that are produced by helm, and run the install.
+```
+# HELM DEMO
```
Comment (hashhar) [inline _episodes/24.md:368]: Is this heading suposed to be inside the code block or an actual heading?

Comment (bitsondatadev) [reply _episodes/24.md:368]: That's okay this was just to split it in the code portion of the demo but wasn't intended for markdown.

**Pull request merged**